CONTRIBUTING.md

```diff
@@ -13,14 +13,80 @@ Please fill out either the individual or corporate Contributor License Agreement
+
+If you have several commits in a pull request and have been asked to squash your commits, please use ```git reset --soft HEAD~N_COMMITS``` and commit again to make your PR a single commit.
+
+### Windows Developers
+
+The symbolic links contained in this repo does not work for Windows operating systems. If you are a Windows developer, please run the [fix](scripts/windows-setup-fix.bat) inside the scripts folder or manually copy the content of the [kubernetes/base](https://github.com/kubernetes-client/python-base) folder into the [kubernetes](kubernetes) folder.
+
+### Writing Tests
+
+In addition to running the your fix yourself and tell us that your fix works, you can demonstrate that your fix really works by using unit tests and end to end tests. These unit tests are mainly located in three places. You should put your tests into the places that they fit in.
+
+1. [Generated tests](kubernetes/test) by OpenAPI generator: these tests should pass and do not require modification.
+2. [End to end tests](kubernetes/e2e_test): these are tests that can only be verified with a live kubernetes server.
+3. Base repo tests in the [base](https://github.com/kubernetes-client/python-base) repo, in which the test files are named ```test_*.py```: These tests use the package ```Mock``` and confirms the functionality of the base repo files.
+
+### Coding Style
+
+We use an automatic coding style checker by using the ```diff``` of the autopep8 output and your code file. To make sure that your code passes the coding style checker, run ```autopep8 --in-place --aggressive --aggressive your_code.py``` before committing and submission.
+
+## Running Tests Locally
+
+If you write a new end to end (e2e) test, or change behaviors that affect e2e tests, you should set up a local cluster and test them on your machine. The following steps will help you run the unit tests.
+
+1. Acquire a local cluster. [Minikube](https://github.com/kubernetes/minikube) is a good choice for Windows and Linux developers. Alternatively if you are on Linux, you can clone the [kubernetes](https://github.com/kubernetes/kubernetes) and run [install-etcd.sh](https://github.com/kubernetes/kubernetes/blob/master/hack/install-etcd.sh) and then [local-up-cluster.sh](https://github.com/kubernetes/kubernetes/blob/master/hack/local-up-cluster.sh) to get a local cluster up and running.
+
+2. Run the unit tests. In the root directory of the main repo, run ```python -m unittest discover```.
+
+3. Check the test results and make corresponding fixes.
+
+## Update the Base Submodule in the Main Repo After Your python-base PR Is Merged
+
+Your contribution of the base repo will not be automatically reflected in the main repo after your PR is merged. Instead, please update the ```base``` submodule in your fork of the main repo as follows:
+```bash
+$ cd kubernetes/base
+$ git fetch origin
+$ git checkout origin/master # assuming your change has been merged into the latest python-base MASTER
+$ cd ../..
```

Comment:
I was just wondering if we can do the same using `git submodule update --remote`?

Reply:
I've been using this method and had no problem with it as of yet.

Reply:
Yes. The submodule way is a one-shot method to do the same.

Reply:
Wow that's way better than what I was doing. Now I learned some nice trick :) Thanks